deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.kt

```diff
@@ -0,0 +1,465 @@
+            putExtra(DeepLinkHandler.EXTRA_URI, uri?.toString() ?: "")
+            putExtra(DeepLinkHandler.EXTRA_URI_TEMPLATE, uriTemplate ?: "")
+            putExtra(DeepLinkHandler.EXTRA_SUCCESSFUL, !isError)
+            if (isError) {
+                putExtra(DeepLinkHandler.EXTRA_ERROR_MESSAGE, errorMessage)
+            }
+        }
+        LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
+    }
+
+    private fun dispatchResult(
+        result: DeepLinkResult,
+        activity: Activity
+    ) {
+        result.deepLinkMatchResult?.let { matchedDeepLinkResult ->
+            when (matchedDeepLinkResult.deeplinkEntry) {
+                is DeepLinkEntry.MethodDeeplinkEntry ->
+                    result.methodResult.taskStackBuilder?.startActivities()
+                        ?: activity.startActivity(result.methodResult.intent)
+                is DeepLinkEntry.ActivityDeeplinkEntry ->
+                    activity.startActivity(result.methodResult.intent)
+                is DeepLinkEntry.HandlerDeepLinkEntry ->
+                    if (result.isSuccessful) callDeeplinkHandler(result)
+            }
+        }
+    }
+
+    private fun callDeeplinkHandler(result: DeepLinkResult) {
+        val handlerClazz = result.deepLinkMatchResult?.deeplinkEntry?.clazz!!
+        val handlerMethod = handlerClazz.methods.singleOrNull { method ->
+            method.name == DEEP_LINK_HANDLER_METHOD_NAME && method.genericParameterTypes.let { parameterTypes ->
+                parameterTypes.singleOrNull()?.let { it != Object::class.java } ?: false
+            }
+        } ?: error("DeepLinkHandler class ${handlerClazz.name} has more than one methods " +
+                "\"$DEEP_LINK_HANDLER_METHOD_NAME\" methods with a single parameter.")
+        // Ok to call single here as we would have failed above if we would have more than one type.
+        val handlerParameterClazz = handlerMethod.genericParameterTypes.single() as Class<*>
+        val handlerParameterClazzConstructor = handlerParameterClazz.constructors.singleOrNull()
+            ?: error("Handler parameter class can only have one constructor.")
+        val annotationList: List<DeeplinkParam> =
+            handlerParameterClazzConstructor.parameterAnnotations.flatten()
+                .filterNotNull().filter { it.annotationClass == DeeplinkParam::class }
+                .map { it as DeeplinkParam }
+        val typeNameMap = handlerParameterClazzConstructor.parameterTypes.let {
+            // Check if we have as many parameters as annotations, the DeeplinkParam annotation is not
+            // repeatable we cannot have multiple for just one element.
+            if (annotationList.size == it.size) {
+                annotationList.zip(it)
+            } else error("There are ${annotationList.size} annotations but ${it.size} parameters!")
+        }
+
+        val handlerInstance = try {
+            handlerClazz.getField("INSTANCE").get(null)
```

Comment:
Do we need to update any project keep rules?

Reply:
We do need to update them, good catch. I thought about it but then forgot.